File: iosApp/iosApp/DemoThemeSource.swift
//
//  DemoThemeSource.swift
//  PromotionSDKDemo
//
//  Hai cách host lấy một PromotionSDKTheme — đối ứng `DemoThemeSource.kt` bên Android.
//
//  - fromJsonFile(): đọc `promotion_theme.json` trong bundle (CÙNG MỘT FILE với bản ở
//    androidApp/src/main/assets) rồi parse bằng PromotionThemeJson. Trả lời cho câu hỏi "SDK dùng
//    file JSON được không": được, một file dùng chung hai nền tảng.
//  - fromObject(): dựng token bằng code, không qua JSON.
//
//  Cả hai cho cùng bộ màu (teal #2CA196) để thấy hai đường đi ra cùng kết quả.
//
//  File này import PRM nên CỐ TÌNH GIỮ NHỎ — xem ghi chú ở ThemePreviewViewController.
//

import UIKit
import PRM

enum DemoThemeSource {

    static let resourceName = "promotion_theme"

    /// `nil` nếu thiếu file hoặc JSON hỏng — host tự quyết định giữ theme cũ.
    static func fromJsonFile() -> PromotionSDKTheme? {
        guard let url = Bundle.main.url(forResource: resourceName, withExtension: "json"),
              let json = try? String(contentsOf: url, encoding: .utf8) else { return nil }
        return PromotionThemeJson.fromJson(json)
    }

    /// Cùng bộ màu với `fromJsonFile()`, nhưng viết thẳng bằng token — không đụng JSON.
    static func fromObject() -> PromotionSDKTheme {
        let teal = UIColor(red: 0x2C / 255, green: 0xA1 / 255, blue: 0x96 / 255, alpha: 1)
        let tealLight = UIColor(red: 0xEA / 255, green: 0xF6 / 255, blue: 0xF4 / 255, alpha: 1)
        let white = UIColor.white
        let grey = UIColor(white: 0x7A / 255, alpha: 1)
        let greyLight = UIColor(white: 0xF4 / 255, alpha: 1)
        let greyMid = UIColor(white: 0xA7 / 255, alpha: 1)

        return PromotionSDKTheme(
            buttonToken: ButtonToken(
                backgroundColor: teal,
                textColor: white,
                shadowColor: teal.withAlphaComponent(0.2),
                cornerRadius: 12
            ),
            searchBarToken: SearchBarToken(
                borderColor: teal,
                hintTextColor: grey,
                textColor: UIColor(white: 0x22 / 255, alpha: 1),
                iconColor: teal,
                cornerRadius: 10
            ),
            listItemToken: ListItemToken(
                linkTextColor: teal,
                usedBadgeTextColor: grey,
                usedBadgeBackgroundColor: greyLight,
                radioButtonStrokeColor: greyMid,
                radioButtonSelectedStrokeColor: teal
            ),
            tabChipToken: TabChipToken(
                activeBackgroundColor: teal,
                inactiveBackgroundColor: tealLight,
                activeTextColor: white,
                inactiveTextColor: UIColor(white: 0x4E / 255, alpha: 1),
                cornerRadius: 16
            ),
            tabUnderlineToken: TabUnderlineToken(
                indicatorColor: teal,
                activeTextColor: teal,
                inactiveTextColor: grey,
                backgroundColor: white
            ),
            discountBadgeToken: DiscountBadgeToken(
                availableTextColor: teal,
                unavailableTextColor: greyMid,
                availableBackgroundColor: tealLight,
                unavailableBackgroundColor: greyLight,
                actionTextColor: teal
            )
        )
    }
}

File: androidApp/src/main/java/com/ttcn/promotionsdk/app/theme/DemoThemeSource.kt
package com.ttcn.promotionsdk.app.theme

import android.content.Context
import android.graphics.Color
import com.ttcn.prm.ui.theme.PromotionSDKTheme
import com.ttcn.prm.ui.theme.PromotionThemeJson
import com.ttcn.prm.ui.theme.token.PRMButtonToken
import com.ttcn.prm.ui.theme.token.PRMDiscountBadgeToken
import com.ttcn.prm.ui.theme.token.PRMListItemToken
import com.ttcn.prm.ui.theme.token.PRMSearchBarToken
import com.ttcn.prm.ui.theme.token.PRMTabChipToken
import com.ttcn.prm.ui.theme.token.PRMTabUnderlineToken

/**
 * Hai cách host lấy một [PromotionSDKTheme] — đối ứng `DemoThemeSource.swift` bên iOS.
 *
 * - [fromJsonFile]: đọc `assets/promotion_theme.json` (**cùng một file** với bản trong bundle iOS)
 *   rồi parse bằng [PromotionThemeJson]. Đây là câu trả lời cho "SDK dùng file JSON được không": được,
 *   một file dùng chung hai nền tảng.
 * - [fromObject]: dựng token bằng code, không qua JSON.
 *
 * Cả hai trả về cùng một bộ màu (teal `#2CA196`) để so sánh hai đường đi cho ra kết quả giống nhau.
 */
object DemoThemeSource {

    const val ASSET_NAME = "promotion_theme.json"

    /** `null` nếu file thiếu hoặc JSON hỏng — host tự quyết định giữ theme cũ. */
    fun fromJsonFile(context: Context): PromotionSDKTheme? =
        runCatching { context.assets.open(ASSET_NAME).bufferedReader().use { it.readText() } }
            .getOrNull()
            ?.let(PromotionThemeJson::fromJson)

    /** Cùng bộ màu với [fromJsonFile], nhưng viết thẳng bằng token — không đụng JSON. */
    fun fromObject(): PromotionSDKTheme {
        val teal = Color.parseColor("#2CA196")
        val tealLight = Color.parseColor("#EAF6F4")
        val white = Color.parseColor("#FFFFFF")
        val grey = Color.parseColor("#7A7A7A")
        val greyLight = Color.parseColor("#F4F4F4")

        return PromotionSDKTheme(
            buttonToken = PRMButtonToken(
                backgroundColor = teal,
                textColor = white,
                shadowColor = Color.parseColor("#332CA196"),
                cornerRadius = 12f,
            ),
            searchBarToken = PRMSearchBarToken(
                borderColor = teal,
                hintTextColor = grey,
                textColor = Color.parseColor("#222222"),
                iconColor = teal,
                cornerRadius = 10f,
            ),
            listItemToken = PRMListItemToken(
                linkTextColor = teal,
                usedBadgeTextColor = grey,
                usedBadgeBackgroundColor = greyLight,
                radioButtonStrokeColor = Color.parseColor("#A7A7A7"),
                radioButtonSelectedStrokeColor = teal,
            ),
            tabChipToken = PRMTabChipToken(
                activeBackgroundColor = teal,
                inactiveBackgroundColor = tealLight,
                activeTextColor = white,
                inactiveTextColor = Color.parseColor("#4E4E4E"),
                cornerRadius = 16f,
            ),
            tabUnderlineToken = PRMTabUnderlineToken(
                indicatorColor = teal,
                activeTextColor = teal,
                inactiveTextColor = grey,
                backgroundColor = white,
            ),
            discountBadgeToken = PRMDiscountBadgeToken(
                availableTextColor = teal,
                unavailableTextColor = Color.parseColor("#A7A7A7"),
                availableBackgroundColor = tealLight,
                unavailableBackgroundColor = greyLight,
                actionTextColor = teal,
            ),
        )
    }
}
